app/controllers/products_controller.rb

```diff
@@ -37,6 +39,22 @@ def destroy
     end
   end
 
+  def add_to_cart
+    if logged_in?
+      id = params[:id].to_i
+      session[:cart] << id unless session[:cart].include?(id)
+      redirect_to root_path
+    else
+      redirect_to login_path
+    end
+  end
+
+  def remove_from_cart
+    id = params[:id].to_i
```

Comment:
I noticed that you are doing `params[:id].to_i` both here and in `add_to_cart` method. You could extract this into a method something like:

```
def id
 @id ||= params[:id].to_id
end
```

and then just use `id` where you need it